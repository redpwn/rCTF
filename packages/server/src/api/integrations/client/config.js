import clientConfig from '../../../config/client'
import { responses } from '../../../responses'

export default {
  method: 'GET',
  path: '/integrations/client/config',
  requireAuth: false,
  handler: async () => {
    return [responses.goodClientConfig, clientConfig]
  }
}
