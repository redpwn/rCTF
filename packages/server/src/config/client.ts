import { ClientConfig } from './types'
import server from './server'

const config: ClientConfig = {
  meta: server.meta,
  homeContent: server.homeContent,
  sponsors: server.sponsors,
  globalSiteTag: server.globalSiteTag,
  ctfName: server.ctfName,
  divisions: server.divisions,
  defaultDivision: server.defaultDivision,
  origin: server.origin,
  startTime: server.startTime,
  endTime: server.endTime,
  faviconUrl: server.faviconUrl,
  emailEnabled: server.email != null,
  userMembers: server.userMembers,
  ctftime: server.ctftime == null ? undefined : {
    clientId: server.ctftime.clientId
  }
}

export default config
