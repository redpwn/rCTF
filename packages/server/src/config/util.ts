export const parseBoolEnv = (val: string): boolean => {
  return ['true', 'yes', 'y', '1'].includes(val.toLowerCase().trim())
}

export const makeNullsafe = <Arg, Ret>(f: (x: Arg) => Ret): (x: Arg | undefined) => Ret | undefined => {
  return x => (x === undefined) ? undefined : f(x)
}

export const nullsafeParseInt = makeNullsafe(parseInt)
export const nullsafeParseBoolEnv = makeNullsafe(parseBoolEnv)

const _removeUndefined = <T>(o: Record<string, T>): Record<string, T> | undefined => {
  let hasKeys = false
  for (const key of Object.keys(o)) {
    let v = o[key]
    if (typeof v === 'object' && v != null) {
      o[key] = v = _removeUndefined(v as Record<string, unknown>) as T
    }
    if (v === undefined || v === null) {
      delete o[key]
    } else {
      hasKeys = true
    }
  }
  return hasKeys ? o : undefined
}

export const removeUndefined = <T extends Record<string, unknown>>(o: T & Record<string, unknown>): T => {
  const cleaned = _removeUndefined(o) as T | undefined
  return cleaned ?? ({} as T)
}
