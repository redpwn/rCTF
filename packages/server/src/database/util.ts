import { PromiseValue } from 'type-fest'

type ArrayValue<Arr> = Arr extends (infer Val)[] ? Val : Arr

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ExtractQueryType<FuncType extends (...args: any) => any> =
  NonNullable<ArrayValue<PromiseValue<ReturnType<FuncType>>>>
