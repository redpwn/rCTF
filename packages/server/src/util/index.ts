import config from '../config/server'
import clientConfig from '../config/client'
import { promises as fs } from 'fs'
import mustache from 'mustache'
import { FastifyPluginAsync, FastifyRequest, RouteHandlerMethod } from 'fastify'

export * as normalize from './normalize'
export * as validate from './validate'
export * as scores from './scores'
export * as restrict from './restrict'

/**
 * Perform a deep-copy of a JSON-stringifiable object
 */
export const deepCopy = <T>(data: T): T => {
  return JSON.parse(JSON.stringify(data)) as T
}

export const serveIndex: FastifyPluginAsync<{ indexPath: string; }> = async (fastify, opts) => {
  const indexTemplate = (await fs.readFile(opts.indexPath)).toString()

  const rendered = mustache.render(indexTemplate, {
    jsonConfig: JSON.stringify(clientConfig),
    config: clientConfig
  })

  const routeHandler: RouteHandlerMethod = async (req, reply) => {
    void reply.type('text/html; charset=UTF-8')
    void reply.send(rendered)
  }

  fastify.get('/', routeHandler)
  fastify.get('/index.html', async (req, reply) => reply.redirect(301, '/'))
  fastify.get('//*', async (req, reply) => reply.redirect(302, '/'))
  // Fastify bug #2466
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  fastify.setNotFoundHandler(routeHandler)
}

const getFastifyIp = (req: FastifyRequest): string =>
  // Use `get` on req.__proto__ since this function is used in req's getter
  Reflect.get(Object.getPrototypeOf(req), 'ip', req) as typeof req.ip

// Parse Cloudflare CF-Connecting-IP header
const getCloudflareIp = (req: FastifyRequest): string | undefined =>
  req.headers['cf-connecting-ip'] as string | undefined

let getRealIp: (req: FastifyRequest) => string = getFastifyIp

if (config.proxy.cloudflare) {
  getRealIp = (req: FastifyRequest): string =>
    getCloudflareIp(req) || getFastifyIp(req)
}

export { getRealIp }
