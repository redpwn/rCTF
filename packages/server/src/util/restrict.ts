import { ServerConfig } from '../config/types'
import config from '../config/server'

type ACLCheck = (email: string | undefined) => boolean

export interface ACL {
  match: string;
  value: string;
  divisions: (keyof ServerConfig['divisions'])[];
}

interface CompiledACL {
  check: ACLCheck;
  divisions: (keyof ServerConfig['divisions'])[];
}

let acls: CompiledACL[]

const restrictionChecks: { [checkType: string]: (value: string) => ACLCheck } = {
  domain: value => email => email?.endsWith('@' + value) ?? false,
  email: value => email => email === value,
  regex: value => {
    const re = new RegExp(value)
    return email => email === undefined ? false : re.test(email)
  },
  any: value => email => true // eslint-disable-line @typescript-eslint/no-unused-vars
}

export const compileACLs = (): void => {
  let divisionACLs = config.divisionACLs
  // allow everything if no ACLs or if no email verify
  if (!divisionACLs || divisionACLs.length === 0 || !config.email) {
    divisionACLs = [{
      match: 'any',
      value: '',
      divisions: Object.keys(config.divisions)
    }]
  }
  acls = divisionACLs.map(({ match, value, divisions }) => {
    if (!Object.prototype.hasOwnProperty.call(restrictionChecks, match)) {
      throw new Error(`Unrecognized ACL matcher "${match}"`)
    }
    return { check: restrictionChecks[match](value), divisions }
  })
}

compileACLs()

export const allowedDivisions = (email: string | undefined): string[] => {
  for (const acl of acls) {
    if (acl.check(email)) {
      return acl.divisions
    }
  }
  return []
}

export const divisionAllowed = (email: string | undefined, division: string): boolean => {
  return allowedDivisions(email).includes(division)
}
