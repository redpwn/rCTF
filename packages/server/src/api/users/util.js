import * as db from '../../database'
import * as challenges from '../../challenges'
import * as cache from '../../cache'
import { getChallengeInfo } from '../../cache/leaderboard'

export const getGenericUserData = async ({ id }) => {
  const user = await db.users.getUserById({ id })
  if (user === undefined) return null

  return getUserData({ user })
}

export const getUserData = async ({ user }) => {
  let [
    { userSolves, challengeInfo },
    score
  ] = await Promise.all([
    (async () => {
      const userSolves = await db.solves.getSolvesByUserId({ userid: user.id })
      const challengeInfo = await getChallengeInfo({
        ids: userSolves.map((solve) => solve.challengeid)
      })
      return { userSolves, challengeInfo }
    })(),
    cache.leaderboard.getScore({ id: user.id })
  ])

  if (score === null) {
    score = {
      score: 0,
      globalPlace: null,
      divisionPlace: null
    }
  }

  const solves = []

  userSolves.forEach((solve, i) => {
    const chall = challenges.getCleanedChallenge(solve.challengeid)

    // Ignore challenges with invalid id, potentially deleted challs
    if (chall === undefined) return

    solves.push({
      category: chall.category,
      name: chall.name,
      points: challengeInfo[i].score,
      solves: challengeInfo[i].solves,
      id: chall.id,
      createdAt: solve.createdat.valueOf()
    })
  })

  return {
    name: user.name,
    ctftimeId: user.ctftime_id,
    division: user.division,
    score: score.score,
    globalPlace: score.globalPlace,
    divisionPlace: score.divisionPlace,
    solves
  }
}
