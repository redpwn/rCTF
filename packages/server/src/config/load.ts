import { ServerConfig } from './types'
import deepMerge from 'deepmerge'
import { PartialDeep } from 'type-fest'
import { nullsafeParseInt, nullsafeParseBoolEnv, removeUndefined } from './util'
import path from 'path'
import fs from 'fs'
import yaml from 'yaml'

const isRoot = (p: string): boolean => {
  const parsed = path.parse(p)
  return parsed.root === parsed.dir
}

export const findConfigDir = (): string => {
  for (
    let dir = path.resolve(__dirname, '../../');
    !isRoot(dir);
    dir = path.resolve(dir, '../')
  ) {
    const confDir = path.resolve(dir, 'rctf.d')
    if (fs.existsSync(confDir)) {
      return confDir
    }
  }
  throw new Error('Could not infer a config directory')
}

const jsonLoader = (file: string) => JSON.parse(file) as PartialDeep<ServerConfig>
const yamlLoader = (file: string) => yaml.parse(file) as PartialDeep<ServerConfig>

const fileConfigLoaders: Map<string, (file: string) => PartialDeep<ServerConfig>> = new Map([
  ['json', jsonLoader],
  ['yaml', yamlLoader],
  ['yml', yamlLoader]
])

export const loadFile = (file: string): PartialDeep<ServerConfig> | undefined => {
  const matched = /\.([^.]+)$/.exec(file)
  if (matched === null) {
    return
  }
  const loader = fileConfigLoaders.get(matched[1])
  if (loader === undefined) {
    return
  }
  const config = loader(fs.readFileSync(file).toString())
  return config
}

export const loadFileConfigs = (configPath?: string): PartialDeep<ServerConfig>[] => {
  const resolvedConfigPath = configPath ??
    process.env.RCTF_CONF_PATH ??
    findConfigDir()

  return fs.readdirSync(resolvedConfigPath)
    .sort()
    .map(file => path.resolve(resolvedConfigPath, file))
    .map(loadFile)
    .filter((x): x is PartialDeep<ServerConfig> => x !== undefined)
}

export const loadEnvConfig = (): PartialDeep<ServerConfig> => removeUndefined({
  database: {
    sql: process.env.RCTF_DATABASE_URL ?? {
      host: process.env.RCTF_DATABASE_HOST,
      port: nullsafeParseInt(process.env.RCTF_DATABASE_PORT),
      user: process.env.RCTF_DATABASE_USERNAME,
      password: process.env.RCTF_DATABASE_PASSWORD,
      database: process.env.RCTF_DATABASE_DATABASE
    },
    redis: process.env.RCTF_REDIS_URL ?? {
      host: process.env.RCTF_REDIS_HOST,
      port: nullsafeParseInt(process.env.RCTF_REDIS_PORT),
      password: process.env.RCTF_REDIS_PASSWORD,
      database: nullsafeParseInt(process.env.RCTF_REDIS_DATABASE)
    },
    migrate: process.env.RCTF_DATABASE_MIGRATE
  },
  instanceType: process.env.RCTF_INSTANCE_TYPE,
  tokenKey: process.env.RCTF_TOKEN_KEY,
  origin: process.env.RCTF_ORIGIN,
  ctftime: {
    clientId: process.env.RCTF_CTFTIME_CLIENT_ID,
    clientSecret: process.env.RCTF_CTFTIME_CLIENT_SECRET
  },
  userMembers: nullsafeParseBoolEnv(process.env.RCTF_USER_MEMBERS),
  homeContent: process.env.RCTF_HOME_CONTENT,
  ctfName: process.env.RCTF_NAME,
  meta: {
    description: process.env.RCTF_META_DESCRIPTION,
    imageUrl: process.env.RCTF_IMAGE_URL
  },
  faviconUrl: process.env.RCTF_FAVICON_URL,
  globalSiteTag: process.env.RCTF_GLOBAL_SITE_TAG,
  email: {
    from: process.env.RCTF_EMAIL_FROM,
    logoUrl: process.env.RCTF_EMAIL_LOGO_URL
  },
  startTime: nullsafeParseInt(process.env.RCTF_START_TIME),
  endTime: nullsafeParseInt(process.env.RCTF_END_TIME),
  leaderboard: {
    maxLimit: nullsafeParseInt(process.env.RCTF_LEADERBOARD_MAX_LIMIT),
    maxOffset: nullsafeParseInt(process.env.RCTF_LEADERBOARD_MAX_OFFSET),
    updateInterval: nullsafeParseInt(process.env.RCTF_LEADERBOARD_UPDATE_INTERVAL),
    graphMaxTeams: nullsafeParseInt(process.env.RCTF_LEADERBOARD_GRAPH_MAX_TEAMS),
    graphSampleTime: nullsafeParseInt(process.env.RCTF_LEADERBOARD_GRAPH_SAMPLE_TIME)
  },
  loginTimeout: nullsafeParseInt(process.env.RCTF_LOGIN_TIMEOUT)
})

export const defaultConfig: PartialDeep<ServerConfig> = {
  database: {
    migrate: 'never'
  },
  instanceType: 'all',
  userMembers: true,
  sponsors: [],
  homeContent: '',
  faviconUrl: 'https://redpwn.storage.googleapis.com/branding/rctf-favicon.ico',
  challengeProvider: {
    name: 'challenges/database'
  },
  uploadProvider: {
    name: 'uploads/local'
  },
  proxy: {
    cloudflare: false,
    trust: false
  },
  meta: {
    description: '',
    imageUrl: ''
  },
  leaderboard: {
    maxLimit: 100,
    maxOffset: 4294967296,
    updateInterval: 10000,
    graphMaxTeams: 10,
    graphSampleTime: 1800000
  },
  loginTimeout: 3600000
}

export const loadFullServerConfig = (configPath?: string): ServerConfig =>
  deepMerge.all([defaultConfig, ...loadFileConfigs(configPath), loadEnvConfig()]) as ServerConfig
