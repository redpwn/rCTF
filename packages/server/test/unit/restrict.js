import config from '../../src/config/server'
import { divisionAllowed, allowedDivisions, compileACLs } from '../../src/util/restrict'

let old

beforeAll(() => {
  old = { divisionACLs: config.divisionACLs, email: config.email }
  config.email = {}
})

afterAll(() => {
  config.divisionACLs = old.divisionACLs
  config.email = old.email
  compileACLs()
})

test('matches domains', () => {
  config.divisionACLs = [{
    match: 'domain',
    value: 'good-domain.com',
    divisions: ['allowed']
  }]
  compileACLs()
  expect(divisionAllowed('anything@good-domain.com', 'allowed')).toBe(true)
  expect(divisionAllowed('anything@bad-domain.com', 'allowed')).toBe(false)
})

test('matches emails', () => {
  config.divisionACLs = [{
    match: 'email',
    value: 'good-email@test.com',
    divisions: ['allowed']
  }]
  compileACLs()
  expect(divisionAllowed('good-email@test.com', 'allowed')).toBe(true)
  expect(divisionAllowed('bad-email@test.com', 'allowed')).toBe(false)
})

test('matches regex', () => {
  config.divisionACLs = [{
    match: 'regex',
    value: '^regex-email(-[a-z]+)?@test.com$',
    divisions: ['allowed']
  }]
  compileACLs()
  expect(divisionAllowed('regex-email@test.com', 'allowed')).toBe(true)
  expect(divisionAllowed('regex-email-abcd@test.com', 'allowed')).toBe(true)
  expect(divisionAllowed('regex-email@bad.com', 'allowed')).toBe(false)
  expect(divisionAllowed('regex-emailaaa@test.com', 'allowed')).toBe(false)
  expect(divisionAllowed('regex-email-1234@test.com', 'allowed')).toBe(false)
})

test('allows any', () => {
  config.divisionACLs = [{
    match: 'any',
    value: '',
    divisions: ['allowed']
  }]
  compileACLs()
  expect(divisionAllowed('anything@test.com', 'allowed')).toBe(true)
})

test('allows no divisions if no ACL matches', () => {
  config.divisionACLs = [{
    match: 'email',
    value: 'allowed@test.com',
    divisions: ['allowed']
  }]
  compileACLs()
  expect(allowedDivisions('not-allowed@test.com')).toEqual([])
})

test('throws error on invalid matcher', () => {
  config.divisionACLs = [{
    match: 'bad'
  }]
  expect(compileACLs).toThrowErrorMatchingSnapshot('"bad"')

  config.divisionACLs = [{
    match: '__proto__'
  }]
  expect(compileACLs).toThrowErrorMatchingSnapshot('"__proto__"')
})

test('denies no email with all matchers except any', () => {
  config.divisionACLs = [{
    match: 'domain',
    value: 'good-domain.com',
    divisions: ['domain']
  }, {
    match: 'email',
    value: 'allowed@test.com',
    divisions: ['email']
  }, {
    match: 'regex',
    value: '^regex-email(-[a-z]+)?@test.com$',
    divisions: ['regex']
  }, {
    match: 'any',
    value: '',
    divisions: ['any']
  }]
  compileACLs()
  expect(divisionAllowed(undefined, 'domain')).toBe(false)
  expect(divisionAllowed(undefined, 'email')).toBe(false)
  expect(divisionAllowed(undefined, 'regex')).toBe(false)
  expect(divisionAllowed(undefined, 'any')).toBe(true)
})
