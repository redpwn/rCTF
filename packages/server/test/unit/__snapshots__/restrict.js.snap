// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error on invalid matcher: "__proto__" 1`] = `"Unrecognized ACL matcher \\"__proto__\\""`;

exports[`throws error on invalid matcher: "bad" 1`] = `"Unrecognized ACL matcher \\"bad\\""`;
